README.txt
StretchPlayer - Audio file stretch player and phrase trainer
Copyright (C) 2010 by Gabriel M. Beddingfield
All Rights Reserved

"For of him [God], and through him, and unto him, are all things.  To
him be the glory for ever. Amen." (Romans 11:36)

CONTENTS
========
1. Description
2. Dependencies and Installation
3. Usage
4. About StretchPlayer

DESCRIPTION
===========

StretchPlayer is an audio file player that allows you to change the
speed of the song without changing the pitch.  It will also allow you
to transpose the song to another key (while also changing the speed).
This is a very powerful tool for musicians who are learning to play a
pre-recorded song.  It's features include:

    + Time Stretch (50% to 150% of song speed)
    + Pitch shift (up or down 1 octave)
    + A/B repeat

The player supports all the audio formats that libsndfile supports,
which currently includes OGG/Vorbis, WAV, W64, AIFF, SND, and FLAC.

    +--------------------------------------------------+
    | Note that neither libsndfile nor StretchPlayer   |
    | supports MP3 files for patent liability reasons. |
    +--------------------------------------------------+

StretchPlayer is free software, distributed under the GNU Public
License, version 2 or later.  See the files gpl-2.0.txt and
gpl-3.0.txt for the details of the license.

DEPENDENCIES AND INSTALLATION
=============================

Please read the file INSTALL.txt for package dependencies and
installation instructions.

USAGE
=====

$ stretchplayer

If you hover over the buttons with your mouse, you'll find several
keyboard shortcuts.  One extra shortcut is the [HOME] key, which will
set everything back to default settings (including relocating to the
beginning of the song).

    +-------------------------------------------+
    | Important: Performance                    |
    |                                           |
    | Make sure that you use a JACK buffer size |
    | larger than 256.  Less than 256 is known  |
    | to have real-time issues.                 |
    +-------------------------------------------+

ABOUT STRETCHPLAYER
===================

StretchPlayer is free software written to glorify God and His Savior
Jesus Christ.  I dedicate this code to spreading His word and His
love.  If you have found this software useful, please make a donation
to a Christian missionary.  Suggested donation is US $10 or US $20.

The mission is (of course) your choice... but if you don't know one to
pick, may I suggest Wycliffe Bible Translators (www.wycliffe.org).
They are an organization dedicated to translating the Bible into the
"heart language" of every person on earth -- *every* indigenous
language.  In doing so, they have brought literacy to thousands of
third-world peoples.  These peoples now have their language written
down, increased literacy rates, and greater pride in their heritage
through such literacy.  This is admirable even if you are not a
believer of Jesus Christ.

Are you a believer and a follower of Jesus Christ?  Do you have in
your life peace and forgiveness of sins?  Or do you live in fear of
God because of your sins -- the evil things that you do every day?
Jesus didn't come put you down!  He came to SAVE you from your sins.
He came to:

 (1) forgive you for all your sins (past, present, and future)
 (2) Make a way for God to live inside you (every day), to teach you
     about Himself, and to be your friend.

But this can only happen if you surrender your life to Him.  To say,
"I have sinned.  I am sorry.  I turn from that, and turn to you, God.
Please forgive me for my sins.  I give my life to you.  You are now in
control... I am your slave.  I will follow Your Word.  What You say, I
will do."  Not those exact words... but you get the idea.  When this
happens, the Holy Spirit comes into your life and lives within you.
This happens as a "down payment" (from God to you) -- as a way of
proving that you *will* be with God when you die.  Proving that you
are now a child of God -- no longer an enemy but a favored child.
(Otherwise, you will go to hell -- forever separated from God.)

Don't believe me?  OK.  But this is what the Bible (the Christian
Scriptures / Holy Book) says about it:

1. That people are separated from God by sins...
"All have sinned, and fall short of the glory of God." (Romans 3:26)

2. That God loves you and wants to FREE you...
"For God so loved the world, that he gave his only begotten Son, that
whoever believes on him should not perish, but have eternal life. For
God sent not the Son into the world to judge the world; but that the
world should be saved through him. He that believes on him is not
judged: he that does not believe has been judged already, because he
has not believed on the name of the only begotten Son of God." (John
3:16-18)

3. That Jesus is the ONLY way...
"And in no other is there salvation: for there is no other name under
heaven, that is given among men, by which we must be saved." (Acts
4:12)

If you now believe -- tell someone.  Even me.

Offended?  No flames, please.  I'm not forcing myself on anyone.

Questions?  Drop me a line:  gabriel@teuton.org

Grace and peace to you all!



ChangeLog
commit e70cb833b4dd4e69505ac090018fe767fd058730
Merge: ebcfe50 24f8f8c
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Thu Jun 16 23:22:41 2011 -0500

    Merge branch 'features/worker_thread'

commit ebcfe50d4db033f711370da1c1091b1efe58e239
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Thu Jun 16 23:22:13 2011 -0500

    Bump version to 0.504 for development.

M	CMakeLists.txt
M	debian/changelog

commit 24f8f8cd224bcdbcea266b600b95931a72d48495
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Thu Jun 16 22:56:19 2011 -0500

    Engine: Add some documentation in the process() callback.

M	src/Engine.cpp

commit 53895c9acbfaab22f702f29c3304087a52645a94
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Thu Jun 16 22:49:09 2011 -0500

    RubberBandServer: Reformat one block for ledgibility

M	src/RubberBandServer.cpp

commit 53e557c4be048bedaa59dfda7db471fc5dfadf83
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Thu Jun 16 22:42:40 2011 -0500

    Rearrange worker thread a little to put parameter updates at beginning.

M	src/RubberBandServer.cpp

commit 0b9cf9d53262498f6fc786ec8566c3d9652c6f44
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Thu Jun 16 22:11:58 2011 -0500

    RubberBandServer: Improve the thread safety of resets.
    
    This was actually causing the loss-of-first-buffer issue.

M	src/Engine.cpp
M	src/RubberBandServer.cpp

commit bbbeebaedb4856922066e638340ae05ba6749807
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Thu Jun 16 18:59:05 2011 -0500

    RubberBandServer: Fix initial buffer corruption.
    
    On playback, the very first set of samples were becoming corrupt
    because I was pushing data into the stretcher when I shouldn't be.
    Instead of pushing a small amount, call
    RubberBandStretcher::process(bufs,count=0,false).

M	src/RubberBandServer.cpp

commit 458c1b42788033f11816443db975dd2435f84465
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Wed Jun 15 00:45:43 2011 -0500

    Engine: Handle more segment sizes, and adjust if it changes.
    
    If the segment size was large (e.g. 4096) then playing would stall.
    In addition, large segment sizes were not being handled at all because
    of fixed buffer sizes.  This improves some of that.
    
    However, it appears that the first segment or so of input audio is
    being truncated.  This is very apparent for the 4096 size
    segment... but not so much when <= 512.

M	src/AudioSystem.hpp
M	src/Engine.cpp
M	src/Engine.hpp
M	src/JackAudioSystem.cpp
M	src/JackAudioSystem.hpp
M	src/RubberBandServer.cpp
M	src/RubberBandServer.hpp

commit 72abd7a233c4846e2206cceb49cddcdad2a545c6
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Tue Jun 14 22:30:47 2011 -0500

    RubberBandServer: Add CPU load estimator for worker thread.

M	src/Engine.cpp
M	src/RubberBandServer.cpp
M	src/RubberBandServer.hpp

commit c1fa0cb2d87c4b5d0227c51c655f25272ff9847b
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Tue Jun 14 21:45:34 2011 -0500

    Set RT scheduling and use a wait condition on worker thread.

M	src/RubberBandServer.cpp
M	src/RubberBandServer.hpp

commit 8c4918dcff2a241aadebe42454fdf73e33c2e56c
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Tue Jun 14 21:23:34 2011 -0500

    Engine: Remove misc. debugging output.

M	src/Engine.cpp

commit 71d4c522d59087b88b7fdaa4d5ae8dea55dfa11f
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Tue Jun 14 21:17:39 2011 -0500

    Engine: Detect end of song and flush the stretcher buffers
    
    Engine was getting stuck at end of song because there was not enough
    to pull off for a normal cycle in the middle of the song.

M	src/Engine.cpp

commit 057bce2dfc9c6bdea7d39af35db6cdd2dfc8331f
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Tue Jun 14 20:56:59 2011 -0500

    Move buffer feed and latency calcs from Engine to RubberBandServer

M	src/Engine.cpp
M	src/Engine.hpp
M	src/RubberBandServer.cpp
M	src/RubberBandServer.hpp

commit a9fbbe8831f8e2b636348c082d2b6ccda10edef5
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Tue Jun 14 20:42:58 2011 -0500

    Engine.cpp: Fix _n_feed_segs at 2.

M	src/Engine.cpp

commit 93a6ce36103386c4b02fa85cfaaa9fad6d85203d
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Tue Jun 14 20:30:18 2011 -0500

    Improved calculation of loop points and actual output pointer.
    
    Note that the cached number of buffered segments sometimes runs away
    (especially at the end of the song).  However, it's possible that this
    number should be fixed at 2 and controlled by the RubberBandServer.

M	src/Engine.cpp
M	src/Engine.hpp
M	src/RubberBandServer.cpp

commit 9660d8230bffa88ff3c97fc9b1ef11d3544e1aac
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Sun Jun 12 22:56:30 2011 -0500

    AudioSystem: Add interface for frame-based time-stamps.
    
    This is needed to estimate UI input timings.

M	src/AudioSystem.hpp
M	src/JackAudioSystem.cpp
M	src/JackAudioSystem.hpp

commit a1a7403039a5dcf7b15e1d7e771321d343c6cc7b
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Sat Apr 23 13:44:33 2011 -0500

    Add bug about startup error checking.
    
    Reported by Federico Bruni <fedelogy@gmail.com>

M	BUGS.txt

commit 259eac1f5e5bd5f5a13cd718a33018708ee410fd
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Sat Apr 23 13:42:53 2011 -0500

    Commit system to track output position.

M	src/Engine.cpp
M	src/Engine.hpp

commit 123349a02ae6987b2b97d7f7a995c38fed995c04
Author: Gabriel M. Beddingfield <gabriel@teuton.org>
Date:   Sun Feb 6 22:32:00 2011 -0600

    Improve thread safety of A/B repeat.
    
    Instead of changing variables on the fly, increment an atomic integer
    to be processed in the normal process() callback.

M	src/Engine.cpp
M	src/Engine.hpp

commit 63ec35447861ee8096b2fe840f72b6c7d20f1794

src/Engine.cpp
/*
 * Copyright(c) 2009 by Gabriel M. Beddingfield <gabriel@teuton.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY, without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include "Engine.hpp"
#include "AudioSystem.hpp"
#include "RubberBandServer.hpp"
#include "Configuration.hpp"
#include <sndfile.h>
#include <stdexcept>
#include <cassert>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <QFileInfo>
#include <QString>

#include "config.h"

using RubberBand::RubberBandStretcher;

namespace StretchPlayer
{
    Engine::Engine(Configuration *config)
	: _config(config),
	  _playing(false),
	  _hit_end(false),
	  _state_changed(false),
	  _position(0),
	  _loop_a(0),
	  _loop_b(0),
	  _sample_rate(48000.0),
	  _stretch(1.0),
	  _pitch(0),
	  _gain(1.0),
	  _output_position(0)
    {
	QString err;

	QMutexLocker lk(&_audio_lock);

	Configuration::driver_t pref_driver;

	if(_config) {
	    pref_driver = _config->driver();
	}

	_audio_system.reset( audio_system_factory(pref_driver) );

	QString app_name("StretchPlayer");
	_audio_system->init( &app_name, _config, &err );
	_audio_system->set_process_callback(Engine::static_process_callback, this);
	_audio_system->set_segment_size_callback(Engine::static_segment_size_callback, this);

	if( ! err.isNull() ) {
	    char msg[513];
	    strncpy(msg, err.toLocal8Bit().data(), 512); 
	    throw std::runtime_error(msg);
	}


	uint32_t sample_rate = _audio_system->sample_rate();

	_stretcher.reset( new RubberBandServer(sample_rate) );
	_stretcher->set_segment_size( _audio_system->current_segment_size() );
	_stretcher->start();

	if( _audio_system->activate(&err) )
	    throw std::runtime_error(err.toLocal8Bit().data());

    }

    Engine::~Engine()
    {
	QMutexLocker lk(&_audio_lock);

	_stretcher->go_idle();
	_stretcher->shutdown();

	_audio_system->deactivate();
	_audio_system->cleanup();

	callback_seq_t::iterator it;
	QMutexLocker lk_cb(&_callback_lock);
	for( it=_error_callbacks.begin() ; it!=_error_callbacks.end() ; ++it ) {
	    (*it)->_parent = 0;
	}
	for( it=_message_callbacks.begin() ; it!=_message_callbacks.end() ; ++it ) {
	    (*it)->_parent = 0;
	}

	_stretcher->wait();
    }

    void Engine::_zero_buffers(uint32_t nframes)
    {
	// MUTEX MUST ALREADY BE LOCKED

	// Just zero the buffers
	void *buf_L = 0, *buf_R = 0;
	buf_L = _audio_system->output_buffer(0);
	if(buf_L) {
	    memset(buf_L, 0, nframes * sizeof(float));
	}
	buf_R = _audio_system->output_buffer(1);
	if(buf_R) {
	    memset(buf_R, 0, nframes * sizeof(float));
	}
    }

    int Engine::segment_size_callback(uint32_t nframes)
    {
	_stretcher->set_segment_size(nframes);
    }

    int Engine::process_callback(uint32_t nframes)
    {
	bool locked = false;

	if( _loop_ab_pressed > 0 ) {
	    _handle_loop_ab();
	}

	try {
	    locked = _audio_lock.tryLock();
	    if(_state_changed) {
		_state_changed = false;
		_stretcher->reset();
		float left[64], right[64];
		while( _stretcher->available_read() > 0 )
		    _stretcher->read_audio(left, right, 64);
		assert( 0 == _stretcher->available_read() );
		_position = _output_position;
	    }
	    if(locked) {
		if(_playing) {
		    if(_left.size()) {
			_process_playing(nframes);
		    } else {
			_playing = false;
		    }
		} else {
		    _zero_buffers(nframes);
		}
	    } else {
		_zero_buffers(nframes);
	    }
	} catch (...) {
	}

	if(locked) _audio_lock.unlock();

	return 0;
    }

    static void apply_gain_to_buffer(float *buf, uint32_t frames, float gain);

    void Engine::_process_playing(uint32_t nframes)
    {
	// MUTEX MUST ALREADY BE LOCKED
	float *buf_L = 0, *buf_R = 0;

	buf_L = _audio_system->output_buffer(0);
	buf_R = _audio_system->output_buffer(1);

	uint32_t srate = _audio_system->sample_rate();
	float time_ratio = srate / _sample_rate / _stretch;

	_stretcher->time_ratio( time_ratio );
	_stretcher->pitch_scale( ::pow(2.0, double(_pitch)/12.0) * _sample_rate / srate );

	uint32_t frame;
	uint32_t reqd, gend, zeros, feed;

	assert( _stretcher->is_running() );

	// Determine how much data to push into the stretcher
	int32_t write_space, written, input_frames;
	write_space = _stretcher->available_write();
	written = _stretcher->written();
	if(written < _stretcher->feed_block_min()
	   && write_space >= _stretcher->feed_block_max() ) {
	    input_frames = _stretcher->feed_block_max();
	} else {
	    input_frames = 0;
	}

	// Push data into the stretcher, observing A/B loop points
	while( input_frames > 0 ) {
	    feed = input_frames;
	    if( looping() && ((_position + feed) >= _loop_b) ) {
		if( _position >= _loop_b ) {
		    _position = _loop_a;
		    if( _loop_a + feed > _loop_b ) {
			assert(_loop_b > _loop_a );
			feed = _loop_b - _loop_a;
		    }
		} else {
		    assert( _loop_b >= _position );
		    feed = _loop_b - _position;
		}
	    }
	    if( _position + feed > _left.size() ) {
		feed = _left.size() - _position;
		input_frames = feed;
	    }
	    _stretcher->write_audio( &_left[_position], &_right[_position], feed );
	    _position += feed;
	    assert( input_frames >= feed );
	    input_frames -= feed;
	    if( looping() && _position >= _loop_b ) {
		_position = _loop_a;
	    }
	}

	// Pull generated data off the stretcher
	uint32_t read_space;
	read_space = _stretcher->available_read();

	if( read_space >= nframes ) {
	    _stretcher->read_audio(buf_L, buf_R, nframes);
	} else if ( (read_space > 0) && _hit_end ) {
	    _zero_buffers(nframes);
	    _stretcher->read_audio(buf_L, buf_R, read_space);
	} else {
	    _zero_buffers(nframes);
	}

	// Update our estimation of the output position.
	unsigned n_feed_buf = _stretcher->latency();
	if(_position > n_feed_buf) {
	    _output_position = _position - n_feed_buf;
	} else {
	    _output_position = 0;
	}
	assert( (_output_position > _position) ? (_output_position - _position) <= n_feed_buf : true );
	assert( (_output_position < _position) ? (_position - _output_position) <= n_feed_buf : true );

	// Apply gain... unroll loop manually so GCC will use SSE
	if(nframes & 0xf) {  // nframes < 16
	    unsigned f = nframes;
	    while(f--) {
		(*buf_L++) *= _gain;
		(*buf_R++) *= _gain;
	    }
	} else {
	    apply_gain_to_buffer(buf_L, nframes, _gain);
	    apply_gain_to_buffer(buf_R, nframes, _gain);
	}

	if(_position >= _left.size()) {
	    _hit_end = true;
	}
	if( (_hit_end == true) && (read_space == 0) ) {
	    _hit_end = false;
	    _playing = false;
	    _position = 0;
	    _stretcher->reset();
	}

	// Wake up, lazybones!
	_stretcher->nudge();
    }

    /**
     * Load a file
     *
     * \return Name of song
     */
    QString Engine::load_song(const QString& filename)
    {
	QMutexLocker lk(&_audio_lock);
	stop();
	_left.clear();
	_right.clear();
	_position = 0;
	_output_position = 0;
	_stretcher->reset();

	SNDFILE *sf = 0;
	SF_INFO sf_info;
	memset(&sf_info, 0, sizeof(sf_info));

	_message( QString("Opening file...") );
	sf = sf_open(filename.toLocal8Bit().data(), SFM_READ, &sf_info);
	if( !sf ) {
	    _error( QString("Error opening file '%1': %2")
		    .arg(filename)
		    .arg( sf_strerror(sf) ) );
	    return QString();
	}

	_sample_rate = sf_info.samplerate;
	_left.reserve( sf_info.frames );
	_right.reserve( sf_info.frames );

	if(sf_info.frames == 0) {
	    _error( QString("Error opening file '%1': File is empty")
		    .arg(filename) );
	    sf_close(sf);
	    return QString();
	}

	_message( QString("Opening file...") );
	std::vector<float> buf(4096, 0.0f);
	sf_count_t read, k;
	unsigned mod;
	while(true) {
	    read = sf_read_float(sf, &buf[0], buf.size());
	    if( read < 1 ) break;
	    for(k=0 ; k<read ; ++k) {
		mod = k % sf_info.channels;
		if( mod == 0 ) {
		    _left.push_back( buf[k] );
		} else if( mod == 1 ) {
		    _right.push_back( buf[k] );
		} else {
		    // remaining channels ignored
		}
	    }
	}

	if( _left.size() != sf_info.frames ) {
	    _error( QString("Warning: not all of the file data was read.") );
	}

	sf_close(sf);
	QFileInfo f_info(filename);
	return f_info.fileName();
    }

    void Engine::play()
    {
	if( ! _playing ) {
	    _state_changed = true;
	    _playing = true;
	}
    }

    void Engine::play_pause()
    {
	_playing = (_playing) ? false : true;
	_state_changed = true;
    }

    void Engine::stop()
    {
	if( _playing ) {
	    _playing = false;
	    _state_changed = true;
	}
    }

    float Engine::get_position()
    {
	if(_left.size() > 0) {
	    return float(_output_position) / _sample_rate;
	}
	return 0;
    }

    void Engine::loop_ab()
    {
	_loop_ab_pressed.fetchAndAddRelaxed(1);
    }

    void Engine::_handle_loop_ab()
    {
	while( _loop_ab_pressed > 0 ) {
	    uint32_t pos, lat;
	    uint32_t pressed_frame, seg_frame;

	    assert( _stretcher->time_ratio() > 0 );
	    pos = _output_position;

	    if(pos > lat) pos -= lat;

	    if( _loop_b > _loop_a ) {
		_loop_b = 0;
		_loop_a = 0;
	    } else if( _loop_a == 0 ) {
		_loop_a = pos;
		if(pos == 0) {
		    _loop_a = 1;
		}
	    } else if( _loop_a != 0 ) {
		if( pos > _loop_a ) {
		    _loop_b = pos;
		} else {
		    _loop_a = pos;
		}
	    } else {
		assert(false);  // invalid state
	    }
	    _loop_ab_pressed.fetchAndAddOrdered(-1);
	}
    }

    float Engine::get_length()
    {
	if(_left.size() > 0) {
	    return float(_left.size()) / _sample_rate;
	}
	return 0;
    }

    void Engine::locate(double secs)
    {
	unsigned long pos = secs * _sample_rate;
	QMutexLocker lk(&_audio_lock);
	_output_position = _position = pos;
	_state_changed = true;
	_stretcher->reset();
    }

    void Engine::_dispatch_message(const Engine::callback_seq_t& seq, const QString& msg) const
    {
	QMutexLocker lk(&_callback_lock);
	Engine::callback_seq_t::const_iterator it;
	for( it=seq.begin() ; it!=seq.end() ; ++it ) {
	    (**it)(msg);
	}
    }

    void Engine::_subscribe_list(Engine::callback_seq_t& seq, EngineMessageCallback* obj)
    {
	if( obj == 0 ) return;
	QMutexLocker lk(&_callback_lock);
	obj->_parent = this;
	seq.insert(obj);
    }

    void Engine::_unsubscribe_list(Engine::callback_seq_t& seq, EngineMessageCallback* obj)
    {
	if( obj == 0 ) return;
	QMutexLocker lk(&_callback_lock);
	obj->_parent = 0;
	seq.erase(obj);
    }

    float Engine::get_cpu_load()
    {
	float audio_load, worker_load;

	audio_load = _audio_system->dsp_load();
	if(_playing) {
	    worker_load = _stretcher->cpu_load();
	} else {
	    worker_load = 0.0;
	}
	return  audio_load + worker_load;
    }

    /* SIMD code for optimizing the gain application.
     *
     * Below is vectorized (SSE, SIMD) code for applying
     * the gain.  If you enable >= SSE2 optimization, then
     * this will calculate 4 floats at a time.  If you do
     * not, then it will still work.
     *
     * This syntax is a GCC extension, but more portable
     * than writing x86 assembly.
     */

    typedef float __vf4 __attribute__((vector_size(16)));
    typedef union {
	float f[4];
	__vf4 v;
    } vf4;

    /**
     * \brief Multiply each element in a buffer by a scalar.
     *
     * For each element in buf[0..nframes-1], buf[i] *= gain.
     *
     * This function detects the buffer alignment, and if it's 4-byte
     * aligned and SSE optimization is enabled, it will use the
     * optimized code-path.  However, it will still work with 1-byte
     * aligned buffers.
     *
     * \param buf - Pointer to a buffer of floats.
     */
    static void apply_gain_to_buffer(float *buf, uint32_t nframes, float gain)
    {
	vf4* opt;
	vf4 gg = {gain, gain, gain, gain};
	int alignment;
	unsigned ctr = nframes/4;

	alignment = (((int)buf)%16);

	switch(alignment) {
	case 4: (*buf++) *= gain;
	case 8: (*buf++) *= gain;
	case 12:(*buf++) *= gain;
	    --ctr;
	case 0:
	    break;
	default:
	  goto LAME;
	}

	assert( (((int)buf)&0xf) == 0 );
	opt = (vf4*) buf;
	while(ctr--) {
	    opt->v *= gg.v; ++opt;
	}

	buf = (float*) opt;
	switch(alignment) {
	case 12: (*buf++) *= gain;
	case 8:  (*buf++) *= gain;
	case 4:  (*buf++) *= gain;
	}

	return;

    LAME:
	// If it's not even 4-byte aligned
	// then this is is the un-optimized code.
	while(nframes--) {
	  (*buf++) *= gain;
	}
    }

} // namespace StretchPlayer
